### DISCUSSION TITLE
What is H2O_WAVE_CHECKPOINT_DIR intended use?

### DISCUSSION THREAD

**Discussion by pascal-pfeiffer** [h2oai/wave#1389] [Q&A]
From the [documentation](https://wave.h2o.ai/docs/configuration#h2o_wave_checkpoint_dir), I read that `H2O_WAVE_CHECKPOINT_DIR` can be used to save contents of `q.app` and `q.user` across restarts of the app for prototyping. 

When trying this feature with the current v0.21 of wave, I see, that `h2o_wave.checkpoint` pickle is only created when changing one or more lines in the code while the wave server is running. This checkpoint is then loaded when restarting the app. In contrast, wave does not save the state of `q.app` and `q.user` when I `ctrl + c` the app server and it shuts down. Is this intended?

**Comment by lo5** [comment]
The checkpointing is tied to the `on_shutdown` callback in Starlette (the underlying framework used in Wave), which means that the checkpoint is created whenever Starlette thinks it's shutting down (typically a sigint or sigterm, but not a sigkill). 

The implementation is fairly simple: https://github.com/h2oai/wave/blob/master/py/h2o_wave/server.py#L388-L400

The situation might be different with the recent changes to `wave run` where it starts both the web server and app server together, which is only intended for development. @mturoci - any ideas?

**Reply by mturoci** [reply]
The Starlette `on_shutdown` hook is not called. (Note: When running the app standalone - with server, it works as intended). According the [docs](https://www.starlette.io/events/):

`The shutdown handlers will run once all connections have been closed, and any in-process background tasks have completed.`

I would say that we are hitting the "any in-process background tasks" which is waved in our case. Will look for some solutions.
